core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/utils/InfraUtils.kt

```diff
@@ -20,3 +17,149 @@ fun TrackNetworkInfra.getNextTrackSections(
+    val res = mutableDirStaticIdxArrayListOf<TrackChunk>()
+    for (block in blockIds) for (chunk in getTrackChunksFromBlock(block)) res.add(chunk)
+    return res
+}
+
+/** Returns all routes that cover the given block */
+fun BlockInfra.routesOnBlock(rawInfra: RawInfra, block: BlockId): StaticIdxList<Route> {
+    val chunks = getTrackChunksFromBlock(block)
+    val routes = rawInfra.getRoutesOnTrackChunk(chunks[0])
+    val res = mutableStaticIdxArrayListOf<Route>()
+    for (routeId in routes) {
+        if (getRouteBlocks(rawInfra, routeId).contains(block)) {
+            res.add(routeId)
+        }
+    }
+    return res
+}
+
+/** Returns the blocks that follow the given one */
+fun BlockInfra.getNextBlocks(infra: RawSignalingInfra, blockId: BlockId): Set<BlockId> {
+    val entry = getBlockExit(infra, blockId)
+    return getBlocksStartingAtDetector(entry).toSet()
+}
+
+/** Returns the block's entry detector */
+fun BlockInfra.getBlockEntry(rawInfra: RawInfra, block: BlockId): DirDetectorId {
+    val blockPath: StaticIdxList<ZonePath> = getBlockPath(block)
+    val firstZone: ZonePathId = blockPath[0]
+    return rawInfra.getZonePathEntry(firstZone)
+}
+
+/** Returns the block's exit detector */
+fun BlockInfra.getBlockExit(rawInfra: RawInfra, block: BlockId): DirDetectorId {
+    val blockPath: StaticIdxList<ZonePath> = getBlockPath(block)
+    val lastZonePath: ZonePathId = blockPath[blockPath.size - 1]
+    return rawInfra.getZonePathExit(lastZonePath)
+}
+
+/** Returns the blocks that lead into the given one */
+fun BlockInfra.getPreviousBlocks(infra: RawSignalingInfra, blockId: BlockId): Set<BlockId> {
+    val entry = getBlockEntry(infra, blockId)
+    return getBlocksEndingAtDetector(entry).toSet()
+}
+
+/** Returns the route's corresponding blocks */
+fun BlockInfra.getRouteBlocks(
+    rawInfra: RawInfra,
+    route: RouteId,
+    allowedSigSystems: StaticIdxList<SignalingSystem>? = null
+): StaticIdxList<Block> {
+    val blockPaths =
+        recoverBlocks(rawInfra, this, mutableStaticIdxArrayListOf(route), allowedSigSystems)
+    if (blockPaths.isEmpty()) return mutableStaticIdxArrayListOf()
+    // No signaling system for now, take the first block path possibility.
+    // Correct when signalisation is taken into account.
+    val blocks = blockPaths[0].toBlockList()
+    return blocks
+}
+
+/** Finds a valid route that follows the given path */
+private fun findRoute(
+    infra: RawSignalingInfra,
+    blockInfra: BlockInfra,
+    chunks: DirStaticIdxList<TrackChunk>,
+    startIndex: Int,
+    routeMustIncludeStart: Boolean
+): RouteId {
+    val routes = infra.getRoutesOnTrackChunk(chunks[startIndex])
+
+    // We need to evaluate the longest route first, in case one route covers a subset of another
+    val sortedRoutes = routes.sortedBy { r -> -infra.getChunksOnRoute(r).size }
+    for (routeId in sortedRoutes) if (
+        routeMatchPath(infra, blockInfra, chunks, startIndex, routeMustIncludeStart, routeId)
+    )
```

Comment:
that's some very cursed formatting, can you add brackets to help the formatter unfuck it?